### PR TITLE
fix: use materialized view for weekly downloads to prevent OOM

### LINKED ISSUE #318: Fix OOM on CH
https://nugettrends.sentry.io/issues/trace/a88e1bec0a444c1b9306788c33705215/?environment=production&groupId=6905208507&node=span-b08fa20319ff2f3a&pageEnd=2025-12-31T04%3A02%3A21.083&pageStart=2025-12-30T04%3A02%3A21.083&project=1266321&referrer=event-or-group-header&source=issue_details&timestamp=1767110541.027

Image

### PR DESCRIPTION
**Pull request by bruno-garcia**
## Summary

Fixes #318

- Create `weekly_downloads` table with `AggregatingMergeTree` engine to store pre-aggregated weekly averages
- Create `weekly_downloads_mv` materialized view that transforms daily inserts into weekly aggregates using `avgState()` at insert time
- Update `GetWeeklyDownloadsAsync` to query the pre-aggregated table using `avgMerge()` - eliminates memory-intensive runtime aggregation
- Add comprehensive tests for varying days per week (1, 2, 3, 5, 7 days) to ensure correct average calculation regardless of data gaps

## Root Cause

The previous query performed `GROUP BY toMonday(date)` with `avg(download_count)` at query time. For packages with 20 years of daily data (~7,300 rows), ClickHouse's in-memory aggregation exceeded the 6.99 GiB memory limit.

## Solution

Pre-aggregate daily downloads into weekly buckets at insert time using a materialized view. Queries now read from the pre-aggregated table with a trivial `avgMerge()` operation that requires minimal memory.

## Deployment Steps

**Important**: This requires manual steps before deploying the code.

1. **Create the MV and table** in production ClickHouse:
   ```sql
   -- Run the CREATE TABLE and CREATE MATERIALIZED VIEW statements from
   -- src/NuGetTrends.Data/ClickHouse/migrations/2026-01-03-01-weekly-downloads-mv.sql
   ```

2. **Run the one-time backfill**:
   ```sql
   INSERT INTO nugettrends.weekly_downloads
   SELECT
       package_id,
       toMonday(date) AS week,
       avgState(download_count) AS download_avg
   FROM nugettrends.daily_downloads
   GROUP BY package_id, week;
   ```

3. **Deploy the code** - the app will now query `weekly_downloads` instead of `daily_downloads`

## Testing

- Added 6 new tests covering edge cases: 1, 2, 3, 5, 7 days per week, non-consecutive days, and mixed weeks with different day counts
- All tests verify correct average calculation regardless of missing days (simulating worker failures)

### CODE REVIEW

File: src/NuGetTrends.Data/ClickHouse/ClickHouseService.cs [hunk 1]
```diff
@@ -164,13 +164,17 @@ public async Task<List<DailyDownloadResult>> GetWeeklyDownloadsAsync(
...
+        // GROUP BY is still needed as ClickHouse may not have merged all parts yet,
+        // but this is now a trivial merge of aggregate states, not full aggregation
         const string query = """
             SELECT
-                toMonday(date) AS week,
-                avg(download_count) AS download_count
-            FROM daily_downloads
+                week,
+                avgMerge(download_avg) AS download_count
+            FROM weekly_downloads
```
[the remark this answers is not in the document]
Comment (bruno-garcia) [reply src/NuGetTrends.Data/ClickHouse/ClickHouseService.cs:175]: Why This Is Correct
The weekly_downloads table stores data aggregated by week, where week is always a Monday (the start of the week). If we used the exact date cutoff without toMonday(), we could:
1. Miss partial weeks: If today() - 3 MONTH falls on a Thursday, we'd miss Mon-Wed of that week, even though the aggregate includes all 7 days.
2. Return inconsistent data: The weekly average for a week would be included or excluded based on an arbitrary day within that week.
The Behavior Change
Yes, this can return up to 6 extra days of data (one partial week). But this is semantically correct because:
- We're returning whole weeks, not partial data
- The API returns weekly aggregates, so returning a partial week would be misleading
- Users requesting "3 months of weekly data" expect complete weeks
Regarding the Test
The AI is correct that GetWeeklyDownloadsAsync_FiltersCorrectDateRange doesn't specifically test this boundary behavior. However:
1. The current tests verify the data returned is correct for the weeks included
2. The "extra" data (up to 6 days) is at most one additional week in the response
